### PR TITLE
Add support for custom fingerprint

### LINKED ISSUE #423: Support for custom fingerprint
Implement custom fingerprint in our java client.

Front logo [Front conversations](https://app.frontapp.com/open/top_gw141)

### PR DESCRIPTION
**Pull request by MaximeAnsquer**
Resolves #423

### CODE REVIEW

File: honeybadger-java/src/main/java/io/honeybadger/reporter/HoneybadgerReporter.java [hunk 1]
```diff
@@ -260,7 +271,7 @@ protected NoticeReportResult submitError(final Throwable error,
             }
 
             NoticeDetails noticeDetails = new NoticeDetails(
-                    getConfig(), error, tags, reportedMessage);
+                    getConfig(), error, tags, reportedMessage, fingerprint);
             notice.setRequest(request).setError(noticeDetails);
         } else {
             NoticeDetails noticeDetails = new NoticeDetails(getConfig(), error, tags);
```
Comment (subzero10) [inline honeybadger-java/src/main/java/io/honeybadger/reporter/HoneybadgerReporter.java:277]: You should pass the `fingerprint` param here as well.

File: honeybadger-java/src/main/java/io/honeybadger/reporter/dto/NoticeDetails.java [hunk 2]
```diff
@@ -26,18 +26,19 @@ public class NoticeDetails implements Serializable {
     private final Set<String> tags;
     private final Backtrace backtrace;
     private final Causes causes;
+    private final String fingerprint;
```
Comment (subzero10) [inline honeybadger-java/src/main/java/io/honeybadger/reporter/dto/NoticeDetails.java:29]: Can you please add `fingerprint` in the `@JsonPropertyOrder` annotation on Line 19 (just for consistency)?

File: honeybadger-java/src/main/java/io/honeybadger/reporter/dto/NoticeDetails.java [hunk 3]
```diff
@@ -47,6 +48,7 @@ public NoticeDetails(final ConfigContext config, final Throwable error, final Se
         this.tags = tags;
         this.backtrace = new Backtrace(config, error);
         this.causes = new Causes(config, error);
+        this.fingerprint = fingerprint;
     }
 
     @Override
```
Comment (subzero10) [inline honeybadger-java/src/main/java/io/honeybadger/reporter/dto/NoticeDetails.java:54]: The `fingerprint` field is missing from the `equals` and `hashCode` methods.

**Review by subzero10 — CHANGES_REQUESTED** [review]
Looks good! Just missing a few things.

**Comment by MaximeAnsquer** [comment]
> Looks good! Just missing a few things.

Right I went a bit too fast, here you go [7383203](https://github.com/honeybadger-io/honeybadger-java/pull/433/commits/7383203582a7896c56edb9e91be08320b05bbc3e)

**Review by subzero10 — APPROVED** [review]

**Comment by MaximeAnsquer** [comment]
@subzero10 any planned date for the release? Thanks. 🤓

**Comment by subzero10** [comment]
Hello @MaximeAnsquer, our release process is quite lengthy and manual, so I'm working on automating it. I should be done before the end of this week. If not, I will release manually and ping you!

**Comment by joshuap** [comment]
2.1.1 has been released. Can you try it out and let us know if it works for you?

**Comment by MaximeAnsquer** [comment]
@joshuap I'm afraid I've introduced a bug.   
When reporting an error, if the `request` parameter is `null` then the reported `message` will always be `null`.
In HoneybadgerReporter.java line 277, I passed `message = null` in the `NoticeDetails` constructor but I should have passed `reportedMessage` or `error.getMessage()`, I'm not sure which one. `reportedMessage` would make more sense to me but it would not be strictly equivalent to what it was before.   
I can submit a PR to fix the bug if you can help me decide on the solution.
Sorry!

**Comment by subzero10** [comment]
> @joshuap I'm afraid I've introduced a bug. When reporting an error, if the `request` parameter is `null` then the reported `message` will always be `null`. In HoneybadgerReporter.java line 277, I passed `message = null` in the `NoticeDetails` constructor but I should have passed `reportedMessage` or `error.getMessage()`, I'm not sure which one. `reportedMessage` would make more sense to me but it would not be strictly equivalent to what it was before. I can submit a PR to fix the bug if you can help me decide on the solution. Sorry!

Hey @MaximeAnsquer, no need to apologise! It was a mistake on my side as well that I didn't see it while reviewing :).
I will fix this and release a new version today!

**Pull request merged**